### ISSUE TITLE
Godot keeps changing ids in theme

### ISSUE THREAD

**Issue by MunWolf** [godotengine/godot#74001]
### Godot version

4.0 rc5

### System information

Windows 10 and Linux

### Issue description

My friend and I are working on a game and we have a theme that we use for all of our UI.

We created it by importing it from the default theme and making some changes (not very important as this was happening before we made the changes)

Every time we save the project if the person who was saving isn't the last one to have commited a change to the theme it changes all the ids on the entries

Example image
![image](https://user-images.githubusercontent.com/3696546/221431550-2a3882d1-bbe2-43aa-b153-34c709fa8116.png)

Notably I am working on the windows release of rc5 while my friend is on the linux release, I assume that has something to do with this issue.

### Steps to reproduce

1. Make a project on the windows version of godot
2. Make a theme
3. Save the project
4. Commit it to git source control
5. Load the project with the linux version of godot
6. Save the project
7. Notice that git status shows changes on the theme, which only includes changes to the resource ids

### Minimal reproduction project

N/A

**Comment by YuriSizov** [comment]
Judging by your screenshot, this happens with the bundled font, and is kind of expected. For the record, you don't have to make a copy of the entire default theme, and at the very least you don't have to bundle the fonts. Themes merge together, and your project theme only needs to contain the parts that you override from the default one.

**Comment by MunWolf** [comment]
That is interesting, because this is what we have in the theme and I don't see any fonts there
![image](https://user-images.githubusercontent.com/3696546/221435169-023cb8d0-a23b-4075-a8ee-f4c0b7a95c66.png)
We initially copied the whole theme and then deleted a bunch of things later on.
Did the font data stick around?

**Comment by YuriSizov** [comment]
Well, I can be wrong, there is only so much I can tell from a screenshot. If you can't provide an MRP, you can try debugging it yourself by recreating the theme anew and adding only the stuff you need.

**Comment by MunWolf** [comment]
Seems like there was 1 font on the label, but it does not show up in the items, only when I try to import from it into another theme.
![image](https://user-images.githubusercontent.com/3696546/221435483-a5812c12-248f-4c2c-9375-59f9e3c5db45.png)

**Comment by MunWolf** [comment]
As an update, including any fonts or textures in your theme from the default theme causes this issue.

**Comment by hsandt** [comment]
I have the same issue, and it seems to affect ImageTexture IDs specifically.

The effect is similar to how a material flagged Local to Scene keeps changing its ShaderMaterial ID (an issue I have yet to report, as I'll try to make a MRP first), but since this is a resource and there is no Local to Scene to check, the cause must be different.

**Comment by dydokamil** [comment]
since godot 4, every change generates 98 line changes (for me) that look like the below diff. This creates quite a mess in the commit history.

```diff
- image = SubResource("Image_ro1df")
+ image = SubResource("Image_sgnnq")

- [sub_resource type="Image" id="Image_702or"]
+ [sub_resource type="Image" id="Image_5qb3y"]
```

**Comment by hsandt** [comment]
I forgot the post the issue I created for Material ID: https://github.com/godotengine/godot/issues/77997
although the cause must be different since it only happens when checking Local to Scene (and this seems to apply to other resources as well)

**Comment by Rindbee** [comment]
https://github.com/godotengine/godot/blob/fc99492d3066098e938449b10e02f8e01d07e2d1/scene/resources/image_texture.cpp#L93

https://github.com/godotengine/godot/blob/fc99492d3066098e938449b10e02f8e01d07e2d1/servers/rendering/renderer_rd/storage_rd/texture_storage.cpp#L796

https://github.com/godotengine/godot/blob/fc99492d3066098e938449b10e02f8e01d07e2d1/servers/rendering/renderer_rd/storage_rd/texture_storage.cpp#L1566-L1567

Setting `Image` for `ImageTexture` ultimately sets a copy of the original `image`. And the copy does not retain the `scene_unique_id` of the original `image`. A new `scene_unique_id` will be generated when the file is saved.

**Comment by Rindbee** [comment]
Similar to #55190.

**Comment by citizenll** [comment]
At the same time, I feel the need to solve a problem. When I copy a node, such as an animation_player, the id of the sub_resource does not change. This can lead to confusing issues. Here is my case: I am currently creating multiple monsters, some of which require AnimationPlayer and Sprite2D nodes. After creating one monster, I want to quickly copy these two nodes to other monster nodes. I expect the nodes to be copied normally. However, the AnimationPlayer node is actually copied with the same id as the resource type, and when I make changes to other monsters, the previous monster is unknowingly affected and damaged.

**Comment by Serotonindude** [comment]
same here... it changes the theme file even if i don't change anything in the theme... got a changed theme file on every commit...

**Comment by 3ter** [comment]
Can someone just shortly reiterate what the current workaround is (except for ignoring the id changes)?

**Comment by jamie-pate** [comment]
> As an update, including any fonts or textures in your theme from the default theme causes this issue.

Remove resources (fonts/textures) from your theme file that reference the default theme or maybe make them unique?

**Comment by 3ter** [comment]
Thanks a ton @jamie-pate ! This button did the trick for me
![image](https://github.com/godotengine/godot/assets/11499929/dc4d6f2b-bd86-4c8e-af24-e367ff0468b2)

**Comment by tinwe** [comment]
We have the same issue and it's driving me crazy that default_theme.tres keeps changing just because it decides to change IDs, so that file gets commited in Github every single commit.

**Comment by DanielKinsman** [comment]
Here is my absurd workaround that possibly might have bad consequences for the integrity of your project:

1. put the theme tres in a subdirectory all of its own
2. remove permissions from that directory and the tres file so you can't write to it (e.g. chown & chmod to root)

Godot will complain about not being able to write to the file whenever you save, but it appears to work...

**Comment by tinwe** [comment]
> Here is my absurd workaround that possibly might have bad consequences for the integrity of your project:
> 
>     1. put the theme tres in a subdirectory all of its own
> 
>     2. remove permissions from that directory and the tres file so you can't write to it (e.g. chown & chmod to root)
> 
> 
> Godot will complain about not being able to write to the file whenever you save, but it appears to work...

Thanks for the suggestions, but that's not really a viable workaround for us, and I find it rather absurd. Has there been any official word about this bug yet?

**Comment by KoBeWi** [comment]
The ID change because you have ImageTexture embedded in your Theme. Trying to get Image of ImageTexture will always return a new instance (#25209), hence the IDs will change.

We already have a warning when importing theme icons:
![image](https://github.com/user-attachments/assets/6ecb3466-820d-480f-98f5-aa7383cf3914)
I guess it could clarify what does it mean exactly or maybe it should be outright not allowed, because icons imported from default/editor theme get embedded. The only way to "fix" this is to save each icon to a new file, so it gets referenced as an external dependency.

**Comment by myyk** [comment]
I think the warning doesn't really let you know how it's not just bigger
but it's unstable and will change frequently which makes it practically
incompatible with source control .

**Comment by DDarby-Lewis** [comment]
This isn't an issue with themes I don't think - I am using ImageTexture in my project and every time the project loads the internal id of the sub-resource containing the image data is changed.

I have tried everything I can think of and everything recommended on other threads to try and fix this but nothing works.
I have even written a system that saves the image as a PNG before loading it into the ImageTexture, hoping that that would sort it but no luck.

This the the kind of ID that keeps changing in the ImageTexture:
```
[sub_resource type="Image" id="Image_spgdd"]
data = {
...
}

image = SubResource("Image_spgdd")

```

If anyone has any ideas I'd be very grateful.

**Comment by myyk** [comment]
For me, (I think) it was that I had to recreate my theme without starting from a copy of the default theme.

**Comment by KoBeWi** [comment]
Save the Image as file instead of embeddimg it. From the inspector, not using script.

**Comment by basilicon** [comment]
Yes, this is an issue with images being embedded in the theme.  This can happen when you don't set a texture for a type; it will load the default texture into your theme, and directly embed the data into your theme as a SubResource.  To fix this: 

1. Look at your theme changes in Git and find the id of the image that is changing.
![image](https://github.com/user-attachments/assets/b0d30f5a-3804-4f3b-99ea-2fd774b7eb67)
2. Open your theme in your preferred text editor, and find where the image is being used (for me it is an ImageTexture).
![image](https://github.com/user-attachments/assets/35d11897-0f35-4fd3-ad0e-5aadc9b145cb)
3. For each id, find where else that is being used; take note of what icon it is being used for (for me it is the LineEdit clear icon).
![image](https://github.com/user-attachments/assets/ccc14cfb-c468-46cf-9593-82a76c54a7eb)
4. In the theme editor in Godot, just clear that associated image, and the embedding disappears.  Alternatively, you can right click the image and Save As, if you are using your own custom texture.
![image](https://github.com/user-attachments/assets/4ccccb6a-be5f-4b45-851f-8a7b716c73c9)

This will remove the embedded data for the texture.  When you play the game, it will still use the texture, it just doesn't load it directly from the theme now.  I don't know why this happens, but I think it can occur during some point of setting up a theme for a certain type for the first time, but other than that I genuinely have no idea how to reproduce this issue.

**Comment by 3ter** [comment]
@basilicon Thank you so much for the step by step solution that at last I was able to follow. Your post will be remembered! Now I don't care that much anymore if it's fixed or not, this workaround is sufficient for me.

**Comment by Pjort** [comment]
Does the new 4.4 update with the UID solve this issue?

**Comment by KoBeWi** [comment]
No, it's unrelated.

**Comment by Megalukes** [comment]
Any solution for that? I have this issue too with any type of resource and it's driving my team crazy because we use git all the time.
